config.go

```diff
@@ -317,5 +335,56 @@ func validateOptions(options Options) (errors []string) {
 		}
 	}
 
+	// options.resource
+	if options.Resources != nil {
```

Comment:
As you are probably well aware, this giant if statement is a bit unwieldy.
As a stylistic thing, keeping the happy path as far left as possible is favorable.  This can be accomplished, sometimes, by failing fast when preconditions on a logic branch are not met.  For example, this _might_ help in this situation:
```
if options.Resources == nil {
  return
}
...
```
And extracting logic into separate functions to make that return statement possible.

And as an additional note, the use of pointers may not even be necessary on the options struct.  That would just avoid the problem of having to check for nil constantly entirely.  Go relies heavily on zero values precisely to avoid nil pointers and the requisite defensive programming they engender.

Ah, the price we pay to get a nicely-nested yaml :/

Reply:
Yeah, but I need to make these pointers, so we can distinguish when the value was actually set in yaml. If we rely on default values (for example for `int` fields), then I am not able to find out whether the user wanted to set the value of the field to `0` or if it is just a default value.

And the similar situation with structs, if I am not mistaken, if those are not pointers, then those fields are set in the yaml file as part of serialization, even if they don't have any values set. I prefer to omit them if user doesn't want to set the specific field, just to make the yaml file more pretty :)

Reply:
I think this is a fair concern about the pointers. These fields should not be used with default values, unless we are specific about what those values are (and that they match the existing knative defaults - which seems error prone itself).

Reply:
@lance so is this PR okay from your POV?

Reply:
Ah, yes I understand.  I mistakenly thought this was the final canonical set of settings.  We have to be able to distinguish between "unspecified" (use the default) and "specifically the zero value".

This is why I have never understood the position held by some that "nil is never correct".